ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java

```diff
@@ -468,313 +351,421 @@ private void unregisterQuery(final QueryMetadata query) {
+          createAsSelect.getPartitionByColumn(),
+          true
       );
-      final AbstractStreamCreateStatement statementWithProperties = streamCreateStatement.copyWith(
-          streamCreateStatement.getElements(),
-          newProperties);
-      return StatementWithSchema.forStatement(
-          statementWithProperties,
-          SqlFormatter.formatSql(statementWithProperties),
-          serviceContext.getSchemaRegistryClient()
+
+      return new PreparedStatement<>(statement.getStatementText(), query);
+    }
+
+    private static PreparedStatement<?> postProcessInsertIntoStatement(
+        final PreparedStatement<?> statement
+    ) {
+      final InsertInto insertInto = (InsertInto) statement.getStatement();
+
+      final QuerySpecification querySpecification =
+          (QuerySpecification) insertInto.getQuery().getQueryBody();
+
+      final Query query = addInto(
+          querySpecification,
+          insertInto.getTarget().getSuffix(),
+          insertInto.getQuery().getLimit(),
+          new HashMap<>(),
+          insertInto.getPartitionByColumn(),
+          false
       );
+
+      return new PreparedStatement<>(statement.getStatementText(), query);
     }
 
-    return StatementWithSchema.forStatement(
-        streamCreateStatement,
-        statementText,
-        serviceContext.getSchemaRegistryClient());
-  }
+    private void validateQuery(final QueryMetadata query, final PreparedStatement<?> statement) {
+      if (statement.getStatement() instanceof CreateStreamAsSelect
+          && query.getDataSourceType() == DataSourceType.KTABLE) {
+        throw new KsqlStatementException("Invalid result type. "
+            + "Your SELECT query produces a TABLE. "
+            + "Please use CREATE TABLE AS SELECT statement instead.",
+            statement.getStatementText());
+      }
 
-  private static void throwOnImmutableOverride(final Map<String, Object> overriddenProperties) {
-    final String immutableProps = overriddenProperties.keySet().stream()
-        .filter(IMMUTABLE_PROPERTIES::contains)
-        .distinct()
-        .collect(Collectors.joining(","));
+      if (statement.getStatement() instanceof CreateTableAsSelect
+          && query.getDataSourceType() == DataSourceType.KSTREAM) {
+        throw new KsqlStatementException("Invalid result type. "
+            + "Your SELECT query produces a STREAM. "
+            + "Please use CREATE STREAM AS SELECT statement instead.",
+            statement.getStatementText());
+      }
 
-    if (!immutableProps.isEmpty()) {
-      throw new IllegalArgumentException("Cannot override properties: " + immutableProps);
+      if (query instanceof PersistentQueryMetadata) {
+        final PersistentQueryMetadata persistentQuery = (PersistentQueryMetadata) query;
+        final SchemaRegistryClient srClient = executionContext.serviceContext
+            .getSchemaRegistryClient();
+
+        if (!AvroUtil.isValidSchemaEvolution(persistentQuery, srClient)) {
```

Comment:
The `KsqlException` is thrown due to errors communicating with the schema store or unexpected errors from the schema store, i.e. 'server errors'.  Most likely the statement is totally fine and either the SR is not reachable or is having issues, e.g. it can't talk to Kafka, etc.

So, these aren't really relating to the statement. The statement itself is likely totally valid. Hence no `KsqlStatementException`, which suggests an issue with the statement.

Reply:
Sure - the only issue i have with it is that the `KsqlException` doesn't have the additional context that may make debugging etc a bit easier. But maybe that is logged somewhere else?

Reply:
Agreed. Not sure on the additional context. However, any change would be outside the scope of this PR me thinks.